Revision: 34480

GriffinLvv opened the pull request:
fix bug: 3282 执行sdbdpsdump时，工具core

【重现步骤】 
1.创建普通表，插入数据 
2.执行sdbdpsdump导出日志信息，检查结果 
【预期结果】 
命令执行成功，且导出日志信息正确 
【实际结果】 
命令执行失败，且工具core： 
```
root@sdbserver2:/opt# /opt/sequoiadb/bin/sdbdpsdump -s /opt/sequoiadb/database/data/11930/replicalog/sequoiadbLog.0 out.log -t 1 
Illegal instruction (core dumped) 
```

core信息如下： 
```
warning: Can't read pathname for load map: Input/output error. 
[Thread debugging using libthread_db enabled] 
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1". 
Core was generated by `/opt/sequoiadb/bin/sdbdpsdump -s /opt/sequoiadb/database/data/11930/replicalog/'. 
Program terminated with signal 4, Illegal instruction. 
#0 0x00000000005969a1 in _dpsDumper::_getSortedFileMap (this=0x7fffe124cb00, 
    dirPath=0x7fffe124dba1 "/opt/sequoiadb/database/data/11930/replicalog/sequoiadbLog.0", mapFiles=...) 
    at SequoiaDB/engine/tools/sdbDpsDump.cpp:1043 
1043	SequoiaDB/engine/tools/sdbDpsDump.cpp: No such file or directory. 
(gdb) bt 
#0 0x00000000005969a1 in _dpsDumper::_getSortedFileMap (this=0x7fffe124cb00, 
    dirPath=0x7fffe124dba1 "/opt/sequoiadb/database/data/11930/replicalog/sequoiadbLog.0", mapFiles=...) 
    at SequoiaDB/engine/tools/sdbDpsDump.cpp:1043 
#1 0x000000000059b9d4 in _dpsDumper::dump (this=0x7fffe124cb00) at SequoiaDB/engine/tools/sdbDpsDump.cpp:750 
#2 0x000000000059cc63 in _dpsDumper::process (this=0x7fffe124cb00, desc=..., vm=...) at SequoiaDB/engine/tools/sdbDpsDump.cpp:652 
#3 0x000000000059cf8d in main (argc=6, argv=0x7fffe124fce8) at SequoiaDB/engine/tools/sdbDpsDump.cpp:1961 
```
【备注】 
1.命令指定/不指定类型-t 都会core掉 
2.replicalog日志和sdbdpsdump工具备份路径：192.168.31.41：/opt/sdbdpsdump

DebbieMadisonS commented:
请安排分析

PhillipsPaull commented:
问题原因分析： 

core的直接原因为下面代码 
```
      else 
      { 
         PD_LOG( PDERROR, "Enum directory[%s] failed, errno: %d", 
                 dirPath, e.code() ) ; 
         rc = SDB_IO ; 
      } 
```
打印时调的函数 e.code() 不为 int 类型，正确应为 e.code().value()。 



同时还存在问题，当输入的路径为文件时，代码 
```
      if ( !fs::exists ( dbDir )) 

      if (!fs::is_directory ( dbDir ) ) 
      { 
         putMatchFileIntoMap(mapFiles, dbDir); 
         goto done ; 
      } 
```
存在逻辑错误。

GriffinLvv commented:
代码提交： 

Revision: 34480 
Date: 2018年2月27日 21:02:50 
Message: 
fix bug: 3282 执行sdbdpsdump时，工具core 
Modified : /trunk/SequoiaDB/engine/tools/sdbDpsDump.cpp 
Modified : /trunk/SequoiaDB/engine/tools/sdbDpsDump.hpp 

Revision: 34481 
Date: 2018年2月27日 21:11:47 
Message: 
fix bug: 3282 执行sdbdpsdump时，工具core 
Modified : /trunk/SequoiaDB/engine/tools/sdbDpsDump.cpp

Revision: 34482 
Date: 2018年2月27日 22:10:00 
Message: 
fix bug: 3282 执行sdbdpsdump时，工具core 
Modified : /trunk/SequoiaDB/engine/include/pmdOptionsMgr.hpp 
Modified : /trunk/SequoiaDB/engine/pmd/pmdOptionsMgr.cpp 
Modified : /trunk/SequoiaDB/engine/tools/sdbDpsDump.cpp 

同步部分优化代码至 2.8分支 

Revision: 34483 
Date: 2018年2月27日 22:17:43 
Message: 
fix bug: 3282 执行sdbdpsdump时，工具core 
同步部分优化代码 
Modified : /branches/sequoiadb_2.8/SequoiaDB/engine/include/pmdOptionsMgr.hpp 
Modified : /branches/sequoiadb_2.8/SequoiaDB/engine/pmd/pmdOptionsMgr.cpp 
Modified : /branches/sequoiadb_2.8/SequoiaDB/engine/tools/sdbDpsDump.cpp 

Revision: 34484 
Date: 2018年2月27日 23:28:37 
Message: 
fix bug: 3282 执行sdbdpsdump时，工具core 
Modified : /trunk/SequoiaDB/engine/tools/sdbDpsDump.cpp 

Revision: 34485 
Date: 2018年2月27日 23:56:27 
Message: 
fix bug: 3282 执行sdbdpsdump时，工具core 
Modified : /trunk/SequoiaDB/engine/tools/sdbDpsDump.cpp

GriffinLvv commented:
检查项---------------------------------------------------------结论（NA/N/Y） 

修改代码是否全部提交SVN主干和分支版本 
且符合代码规范-------------------------------------------- 主干 
是否用例全部通过----------------------------------------- Y 
是否清淅标明重现步骤----------------------------------- Y 
修改代码是否通过检视----------------------------------- N 
资料是否变更---------------------------------------------- NA 
接口是否变更---------------------------------------------- NA 
相关驱动是否变更或添加-------------------------------- NA 
是否保持向前兼容---------------------------------------- Y 
是否影响安装部署和监控-------------------------------- NA 
是否对影响性能------------------------------------------- NA 
sdb shell帮助信息是否更新----------------------------- NA

DebbieMadisonS commented:
请验证

OwenMichaell commented:
【验证结果】验证结果正确 
【验证版本】 
SequoiaDB shell version: 2.9 
Release: 34625 
2018-03-07-00.38.02(Enterprise-Debug) 
【验证步骤】 
1.创建集合，插入数据 
```
> db.createCS('cs').createCL('cl') 
localhost:11810.cs.cl 
Takes 0.473299s. 
> db.cs.cl.insert({a:1}) 
Takes 0.001788s. 
```
2.指定-s为文件路径，执行sdbdpsdump解析日志文件，返回执行成功且解析结果正确： 
`root@sdbserver1:/opt# /opt/sequoiadb/bin/sdbdpsdump -s /opt/sequoiadb/database/data/11930/replicalog/sequoiadbLog.0 -o dpsout.log `
【用例】 
手工用例seqDB-14656:指定-s路径参数为文件，执行sdbdpsdump解析同步日志

请检视

FairyCook commented:
检视通过，请关闭
